jax/core.py

```diff
@@ -1156,85 +1129,100 @@ def check_jaxpr(jaxpr: Jaxpr):
       exception_args = [msg, *e.args[1:]]
     raise exception_type(*exception_args) from e
 
-def _check_jaxpr(jaxpr: Jaxpr):
-  env = _JaxprTypeEnvironment()
-
-  env.write(unitvar)
-  map(env.write, jaxpr.constvars)
-  map(env.write, jaxpr.invars)
+def _check_jaxpr(jaxpr: Jaxpr, in_avals: Sequence[AbstractValue]):
 
-  for eqn in jaxpr.eqns:
-    check_jaxpr_eqn(env, eqn)
+  def read(v: Var) -> AbstractValue:
+    if isinstance(v, Literal):
+      return get_aval(v.val)
+    else:
+      if v not in env:
+        raise TypeError(f"Variable '{v}' not defined")
+      return env[v]
 
-  for subjaxpr in subjaxprs(jaxpr):
-    _check_jaxpr(subjaxpr)
+  def write(v: Var, a: AbstractValue) -> None:
+    if v in env:
+      raise TypeError(f"Variable '{v}' already bound")
+    # TODO(frostig): we'd rather check equality or just typecompat here, but
+    # partial_eval.tracers_to_jaxpr types eqn outvars as abstract_unit if the
+    # outvars are unused
+    if not typecompat(v.aval, a) and v.aval is not abstract_unit:
+      raise TypeError(f"Variable '{v}' inconsistently typed as {a}, "
+                      f"bound as {v.aval}")
+    env[v] = a
 
-  map(env.read, jaxpr.outvars)
+  env : Dict[Var, AbstractValue] = {}
 
-def _valid_eqn_assignment(dst_aval, src_aval):
-  # TODO(frostig): we'd rather this check simply be `typecompat` and not allow
-  # assignment to an AbstractUnit, but partial_eval.tracers_to_jaxpr types eqn
-  # outvars as AbstractUnit if the outvars are unused.
-  return dst_aval is abstract_unit or typecompat(dst_aval, src_aval)
+  write(unitvar, abstract_unit)
+  map(write, jaxpr.constvars, [v.aval for v in jaxpr.constvars])
+  map(write, jaxpr.invars, in_avals)
 
-def check_jaxpr_eqn(env, eqn):
-  invars = map(env.read, eqn.invars)
-  inferred_out_avals = type_transfer(eqn.primitive, invars, eqn.params)
-  outvars = map(env.write, eqn.outvars)
+  for eqn in jaxpr.eqns:
+    in_avals = map(read, eqn.invars)
+    if eqn.primitive.call_primitive:
+      out_avals = check_call(eqn.primitive, in_avals, eqn.params)
+    elif eqn.primitive.map_primitive:
+      out_avals = check_map(eqn.primitive, in_avals, eqn.params)
+    else:
+      out_avals = check_eqn(eqn.primitive, in_avals, eqn.params)
+    try:
+      map(write, eqn.outvars, out_avals)
+    except TypeError as e:
+      msg, = e.args
+      raise TypeError(msg + f" in '{eqn}'") from None
 
-  for outvar, inferred_out_aval in zip(outvars, inferred_out_avals):
-    if not _valid_eqn_assignment(outvar.aval, inferred_out_aval):
-      raise TypeError(
-          f"Jaxpr equation LHS {outvar} is {outvar.aval}, "
-          f"RHS is inferred as {inferred_out_aval}, in '{eqn}'")
+  map(read, jaxpr.outvars)
 
-def type_transfer(prim, invars, params):
-  in_avals = [v.aval for v in invars]
+def check_eqn(prim, in_avals, params):
```

Comment:
@froystig pointed out that to maintain current behavior we need to check any jaxprs in `params` here.

Reply:
Done!